File: 450-delete-node-in-a-bst/450-delete-node-in-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */
var deleteNode = function(root, key) {
  if (root == null) return null
  if (root.val === key) {
    if (!root.left && !root.right) return null
    if (root.left && root.right) {
      let min = root.right
      while (min.left) { min = min.left }
      root.right = deleteNode(root.right, min.val)
      min.left = root.left
      min.right = root.right
      return min
    }
    if (root.left) return root.left
    if (root.right) return root.right
  } 
  if (key < root.val) root.left = deleteNode(root.left, key)
  else root.right = deleteNode(root.right, key)
  return root
};

File: 450-delete-node-in-a-bst/450-delete-node-in-a-bst.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func deleteNode(root *TreeNode, key int) *TreeNode {
  if root == nil { return root }
  if root.Val > key {
    root.Left = deleteNode(root.Left, key)
  }
  if root.Val == key {
    if root.Left == nil && root.Right == nil {
      return nil
    }
    if root.Left == nil {
      return root.Right
    }
    if root.Right == nil {
      return root.Left
    }
    nxt := root.Right
    for nxt.Left != nil {
      nxt = nxt.Left
    }
    root.Right = deleteNode(root.Right, nxt.Val)
    nxt.Left = root.Left
    nxt.Right = root.Right
    return nxt
  }
  root.Right = deleteNode(root.Right, key)
  return root
}